Reduce number of warning logs

pirgeo opened the pull request:
Instead of logging every normalization as warning log, only the first normalization of per program instance will be logged at warning level and all subsequent normalizations will be logged at debug level. This will lead to not immediately flooding users' logs with warning messages if they are using illegal characters.

joaopgrassi commented on lib/src/main/java/com/dynatrace/metric/util/MetricLineBuilderImpl.java:
```diff
@@ -118,8 +137,14 @@ public MetricLineBuilder.TypeStep dimension(String key, String value) throws Met
       if (normalizationResult.messageType() != NormalizationResult.MessageType.NONE) {
-        logger.warning(
-            () -> String.format(PREFIX_STRING, this.metricKey, normalizationResult.getMessage()));
+        Supplier<String> messageSupplier =
```
Couldn't the supplier be re-used and then you pass the parameters to it?

joaopgrassi commented on lib/src/main/java/com/dynatrace/metric/util/MetricLineBuilderImpl.java:
```diff
@@ -81,7 +94,13 @@ public MetricLineBuilder.TypeStep metricKey(String key) throws MetricException {
 
       if (normalizationResult.messageType() == NormalizationResult.MessageType.WARNING) {
-        logger.warning(() -> normalizationResult.getMessage());
+        if (!metricKeyNormalizationWarnLogged) {
```
We could consider wrapping these booleans into a class, and then have specific methods for each case there. So here it would be simpler, something like `NormalizationThrottledLogger.logMetricKeyNormalization()`

pirgeo commented:
I reworked this PR a bit, based on @joaopgrassi's comments.
* Moved the logging into a separate class
* Moved message creation after checking whether or not this line should even be logged. This way, it is possible to get rid of all the Suppliers.

joaopgrassi commented on lib/src/main/java/com/dynatrace/metric/util/NormalizationResult.java:
```diff
@@ -38,7 +38,7 @@ static NormalizationResult newValid(String result) {
   }
 
-  static NormalizationResult newValid(String result, Supplier<String> message) {
+  static NormalizationResult newWarning(String result, Supplier<String> message) {
```
Oops, good find!

joaopgrassi commented on lib/src/main/java/com/dynatrace/metric/util/MetricLinePreConfiguration.java:
```diff
@@ -224,17 +223,11 @@ private void dimension(
+
+        normalizedKey = normalizeKeyResult.getResult();
+        if (normalizeKeyResult.messageType() != NormalizationResult.MessageType.NONE) {
+          normalizationLogger.logDimensionKeyMessage(CLASS_NAME_FOR_LOGGING, normalizeKeyResult);
```
This method asks for metricKey, but here it's passing `CLASS_NAME_FOR_LOGGING`. Is this correct?

pirgeo replied:
Yes, because when we create the pre-configuration, we don't yet know about a metric key. The pre-config contains things like Dynatrace-specific dimensions, and we set it up before we create the first metric builder. In order to make that clear in the logs, I put the class name of the pre-config in, instead of the metric key. `CLASS_NAME_FOR_LOGGING` is basically the class simple name in curly braces. It will lead to a log line like
```
[WARNING] [{MetricLinePreConfiguration}] Dimension key normalized from 'default 1' to 'default_1'. Further normalization logs for data of the same type will be logged at debug level. 
```

joaopgrassi reviewed (commented):
I guess there's no concurrency in place here right? For setting the boolean variables that control the logging?

pirgeo commented:
> I guess there's no concurrency in place here right? For setting the boolean variables that control the logging?

There should not be any concurrency in the current use-cases, but you are right, if used in specific situations, there could be multiple writes at the same time. I'll update it to AtomicBooleans, the overhead should be negligible.

pirgeo commented:
Bumped to `2.0.1`. Should we bump to `2.1.0` or is this fine? It is only a change in the log output, no functional change.

joaopgrassi commented:
I think changing the patch is good enough.
